### PR TITLE
[Core][Proposal-Not merge] Adding new Create method to Condition and Element with parameters

### PR DESCRIPTION
**Pull request by loumalouomega**
I realized that several elements/conditions have custom constructors that are not of easy access in a standard way. For that maybe is a good idea to add a new constructor with Parameters so it can be customized to any needs

### CODE REVIEW

**Comment by pooyan-dadvand** [comment]
I like the idea of having such a constructor, but I'm not sure about the format of it.

**Comment by loumalouomega** [comment]
> I like the idea of having such a constructor, but I'm not sure about the format of it.

The current one is the most basic, just uses the KratosComponents to create the element of that name

**Comment by philbucher** [comment]
we are also discussing it in the @KratosMultiphysics/design-committee and will post once we reached a consensus

**Comment by msandre** [comment]
@KratosMultiphysics/design-committee reached a consensus. We agreed that it may be ok if there is a real need / use case. But we could not identify one where it would make more sense than to use properties. Because it adds more code and dependencies to element.h without a clear use case, we recommend to the @KratosMultiphysics/technical-committee not to proceed with changes for now.

**Comment by loumalouomega** [comment]
> @KratosMultiphysics/design-committee reached a consensus. We agreed that it may be ok if there is a real need / use case. But we could not identify one where it would make more sense than to use properties. Because it adds more code and dependencies to element.h without a clear use case, we recommend to the @KratosMultiphysics/technical-committee not to proceed with changes for now.

The idea is to avoid "not standard" constructors. We have already some elements/conditions with constructors with more arguments than the standard.

**Comment by loumalouomega** [comment]
> > @KratosMultiphysics/design-committee reached a consensus. We agreed that it may be ok if there is a real need / use case. But we could not identify one where it would make more sense than to use properties. Because it adds more code and dependencies to element.h without a clear use case, we recommend to the @KratosMultiphysics/technical-committee not to proceed with changes for now.
> 
> The idea is to avoid "not standard" constructors. We have already some elements/conditions with constructors with more arguments than the standard.

Look at: https://github.com/KratosMultiphysics/Kratos/blob/4c2646dff52a5687fd7e00d45f046b15f307f7de/applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp#L32

**Comment by msandre** [comment]
I will try to summarize the main points that raised concerns in our discussion:
- identification of a clear use case
- added dependencies / compile time
- keeping a unique way of creating elements and conditions in a model part
- memory efficiency: if data is being passed to an element constructor in a Parameters object, then it is probably being passed the same data to many element constructors, which raises the question if it makes sense that every element stores its own copy of the data.

**Comment by loumalouomega** [comment]
> I will try to summarize the main points that raised concerns in our discussion:
> 
>     * identification of a clear use case
> 
I already gave some examples. This is more general and gives more freedom to the developer. We already use this mechanism in the CL.
>     * added dependencies / compile time
> 
I understand the concern. In any case almost everything already depends on Parameters...
>     * keeping a unique way of creating elements and conditions in a model part
> 
I understand the concern. But this way is more general, so we can even discuss remove the simpler modes (passing empty parameters)
>     * memory efficiency: if data is being passed to an element constructor in a Parameters object, then it is probably being passed the same data to many element constructors, which raises the question if it makes sense that every element stores its own copy of the data.

I understand the concern

**Comment by loumalouomega** [comment]
Just to comment that during the Kratos Workshop several people ask me for the feature (or an equivalent feature that allows a flexible constructor to the elements/conditions)

**Comment by RiccardoRossi** [comment]
Just to comment that the "UseRayleighDamping " case is a bad one.

You should NOT use a special constructor for this.

**Comment by loumalouomega** [comment]
> Just to comment that the "UseRayleighDamping " case is a bad one.
> 
> You should NOT use a special constructor for this.

Yes, the idea of this is to avoid this bad procedures

**Comment by RiccardoRossi** [comment]
I have been putting some thinking into this PR, and i have really mixed feelings about this.

To begin i would say that the standard element/condition constructor works pretty nicely and can solve a large number of cases.  

There are surely cases in which one may want to control the behaviour of a given element set. The scenario i am thinking of here is that of a laminated shell, where the user needs to provide the details of the lamination. While it may seem tempting in this context to pass the lamination data through the constructor, i believe that it is much better to control this parameters by passing them through the Properties. 
My rationale in making this statement is that 

1. many elements will most likely share the same data
2. we should avoid passing one parameters object at every object construction since it is slow and memory consuming.
3. we definitely want to avoid STORING a parameters object per ever element (a lot of memory!!, and we shouldn't access it more than once)

There are nevertheless cases in which it may be useful to have multiple elements/conditions to overlap onto the same geometry. I am thinking of something like an "OverlappingElement" which actually combines the action of a few elements that share the same geometry.
For example one might want to construct a weakly coupled Mechanical-thermal element, by combining at construction time two elements into one. 
A constructor in this case could possibly look like

~~~
OverlappingElement(id, geom, prop, 
          """["TotalLagrangian2d4N,
               "ThermalElement2D4N]
         """
            )
~~~      

Such combination at construction time would surely have advantages, unfortunately it also hides some details to be cared about. For example, which GetDofList should the composite element return? how should the RHS and LHS of the combined element look like? (note that the structural element works in DISPLACEMENT and the thermal one in TEMPERATURE) so that the a sort of local assembling would be needed to sum up the element contributions.
In short the point here is that the combined element LHS would need to look something like

~~~
[ [ Kstr   0 ]
 [  0       Kth ]
~~~

while it is definitely possible to make this local assembly to happen automatically, i dare say that the only way of making it fast would be to know at compile time that the element combines a structural and a thermal element. Furthermore a generic element would have no way to tell that the cross terms (the zeros) are anything other than zero.
Now my point is that if we want to make the element fast or to prescribe the combination terms, we would need to have a specialized StructuralThermal element. But if we make a specialized element, than we could keep the standard constructor and simply hard code in there the element combination.

If we believe that the flexibility is completely needed, than we could define a 
"ELEMENT_CONSTRUCTION_PARAMETERS" variable and store it in the properties. This way one could easily access it within the element constructor and ultimately preserve all the flexibility.

An important advantage of this approach over the modified constructor, is that we would avoid futher complicating our mdpa reading (and i strongly believe that we should keep the possibility of reading any element /condition from the mdpa)


Note that my position is less "rigid" regarding the construction of MPCs. In that case WE DO miss a standard construction mechanism... nevertheless i guess that's a different story

**Comment by loumalouomega** [comment]
> I have been putting some thinking into this PR, and i have really mixed feelings about this.
> 
> To begin i would say that the standard element/condition constructor works pretty nicely and can solve a large number of cases.
> 
> There are surely cases in which one may want to control the behaviour of a given element set. The scenario i am thinking of here is that of a laminated shell, where the user needs to provide the details of the lamination. While it may seem tempting in this context to pass the lamination data through the constructor, i believe that it is much better to control this parameters by passing them through the Properties.
> My rationale in making this statement is that
> 
>     1. many elements will most likely share the same data
> 
>     2. we should avoid passing one parameters object at every object construction since it is slow and memory consuming.
> 
>     3. we definitely want to avoid STORING a parameters object per ever element (a lot of memory!!, and we shouldn't access it more than once)
> 
> 
> There are nevertheless cases in which it may be useful to have multiple elements/conditions to overlap onto the same geometry. I am thinking of something like an "OverlappingElement" which actually combines the action of a few elements that share the same geometry.
> For example one might want to construct a weakly coupled Mechanical-thermal element, by combining at construction time two elements into one.
> A constructor in this case could possibly look like
> 
> ```
> OverlappingElement(id, geom, prop, 
>           """["TotalLagrangian2d4N,
>                "ThermalElement2D4N]
>          """
>             )
> ```
> 
> Such combination at construction time would surely have advantages, unfortunately it also hides some details to be cared about. For example, which GetDofList should the composite element return? how should the RHS and LHS of the combined element look like? (note that the structural element works in DISPLACEMENT and the thermal one in TEMPERATURE) so that the a sort of local assembling would be needed to sum up the element contributions.
> In short the point here is that the combined element LHS would need to look something like
> 
> ```
> [ [ Kstr   0 ]
>  [  0       Kth ]
> ```
> 
> while it is definitely possible to make this local assembly to happen automatically, i dare say that the only way of making it fast would be to know at compile time that the element combines a structural and a thermal element. Furthermore a generic element would have no way to tell that the cross terms (the zeros) are anything other than zero.
> Now my point is that if we want to make the element fast or to prescribe the combination terms, we would need to have a specialized StructuralThermal element. But if we make a specialized element, than we could keep the standard constructor and simply hard code in there the element combination.
> 
> If we believe that the flexibility is completely needed, than we could define a
> "ELEMENT_CONSTRUCTION_PARAMETERS" variable and store it in the properties. This way one could easily access it within the element constructor and ultimately preserve all the flexibility.
> 
> An important advantage of this approach over the modified constructor, is that we would avoid futher complicating our mdpa reading (and i strongly believe that we should keep the possibility of reading any element /condition from the mdpa)
> 
> Note that my position is less "rigid" regarding the construction of MPCs. In that case WE DO miss a standard construction mechanism... nevertheless i guess that's a different story

I think that what you propose is not incompatible with this PR. The Create methods can be used only in case this `ELEMENT_CONSTRUCTION_PARAMETERS` (or `CONSTRUCTION_PARAMETERS` so can be used on conditions and MPC too) is defined :)

**Comment by loumalouomega** [comment]
BTW, one other possible application of this can be in case some day we define an application that works similar to FF++, so the strong formulation can be defined on the configuration parameters.

**Comment by loumalouomega** [comment]
@RiccardoRossi can you summarize what's needed in order to proceed with this PR?

**Comment by loumalouomega** [comment]
Ping

**Comment by loumalouomega** [comment]
Ping @KratosMultiphysics/technical-committee

**Comment by loumalouomega** [comment]
Ping @KratosMultiphysics/technical-committee

**Comment by loumalouomega** [comment]
Ping @KratosMultiphysics/technical-committee

**Comment by RiccardoRossi** [comment]
The summary of my position is to close this PR... but this only reflects my point of view not that of @KratosMultiphysics/technical-committee . No decision has been taken so far

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee supports the idea of the @KratosMultiphysics/design-committee and needs a list of reasonable use cases in which this option would be necessary. Please, note that some of the use cases are now doable thanks to the subproperties and some other changes. @loumalouomega please provide it here (again if you have mention it before) but please try to give some very concrete examples.

**Comment by loumalouomega** [comment]
> @KratosMultiphysics/technical-committee supports the idea of the @KratosMultiphysics/design-committee and needs a list of reasonable use cases in which this option would be necessary. Please, note that some of the use cases are now doable thanks to the subproperties and some other changes. @loumalouomega please provide it here (again if you have mention it before) but please try to give some very concrete examples.

I think that I already mentioned an example. The idea is to be able to create elements/conditions in a very general manner. Without needing, for example, storing in the properties many information that may be used only once (the properties which are used only in the construction and not anymore). This is the most clear and practical example, and the one I think may be more determinant to accept this PR. Later I will introduce a more advanced example.

This is basically the same concept considered in the Create in the CL, which allows to construct complex Cl, despite is not used by default for all the CL. 

Additionally this is a very flexible, and I have in mind the following example. This will allow to create generic elements like the ones from Fenics or FreeFEM++, so we can write the strong form and the element could be constructed on the constructor for that specific problem. I accept this is more complex that it may sound, but is a good example what this mechanism will allow.

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee sees strong relation between this feature and generic factories for elements and conditions. So postpones this new feature until we have better vision how would be the signature and requirements of that factory.

Closing for now and we will retake the discussion whenever the factory takes shape.

**Comment by loumalouomega** [comment]
I assume we keep the branch until this is defined?

**Comment by philbucher** [comment]
I am ok with keeping it

**Comment by loumalouomega** [comment]
> I assume we keep the branch until this is defined?

What is the status of this?

**Comment by loumalouomega** [comment]
Maybe with the register we should reopen the discussion of this

**Comment by loumalouomega** [comment]
@KratosMultiphysics/technical-committee do you thinkthis is worth keeping it?
